share/models/change.py

```diff
@@ -272,3 +235,61 @@ def _resolve_change(self):
             else:
                 change[k] = v
         return change
+
+    def _merge_objects(self, from_obj, into_obj):
+        from share.models.base import ShareObject
+
+        # Avoid same_as chains
+        while into_obj.same_as:
+            into_obj = into_obj.same_as
+        for obj in from_obj._meta.concrete_model.objects.filter(same_as=from_obj.id):
+            obj.change = self
+            obj.same_as = into_obj
+            obj.same_as_version = into_obj.version
+            obj.save()
+
+        concrete_model = into_obj._meta.concrete_model
+        assert concrete_model is from_obj._meta.concrete_model
+
+        for field in concrete_model._meta.get_fields():
+            if field.is_relation and not field.many_to_many and issubclass(field.remote_field.model, ShareObject) and hasattr(field, 'field'):
+                # Update incoming foreign keys to point to into_obj.
+                self._merge_fk_field(from_obj, into_obj, field.field)
+            elif not field.is_relation and field.editable and not field.primary_key:
+                # Update scalar field on into_obj. Use value from more recently modified object, ignoring blank values.
+                self._merge_scalar_field(from_obj, into_obj, field)
+
+        into_obj.change = self
+        into_obj.save()
+
+        from_obj.change = self
+        from_obj.same_as = into_obj
+        from_obj.same_as_version = into_obj.version
+        from_obj.save()
+
+        return into_obj
+
+    def _merge_fk_field(self, from_obj, into_obj, field):
+        for obj in field.model.objects.filter(**{field.name: from_obj.id}):
+            try:
+                obj.change = self
+                setattr(obj, field.name, into_obj)
+                with transaction.atomic():
+                    obj.save()
+            except IntegrityError as e:
+                # Look for conflicting keys in error message, e.g. "Key (agent_id, creative_work_id, type)=(868, 115, share.publisher) already exists."
+                # If possible, merge this object into that one.
+                m = re.search('Key \(([^)]+)\)=\(([^)]+)\) already exists.', e.args[0])
+                if m:
+                    keys = dict(zip(m.group(1).split(', '), m.group(2).split(', ')))
+                    conflicting_obj = obj._meta.model.objects.get(**keys)
+                    obj.refresh_from_db()
+                    self._merge_objects(obj, conflicting_obj)
+                else:
+                    raise e
+
+    def _merge_scalar_field(self, from_obj, into_obj, field):
```

Comment:
Not necessary right now, but this would be a good place to throw in a fuzzy match on strings.
https://en.wikipedia.org/wiki/Approximate_string_matching